import os
# config file name
configfile: '../config/config.yaml'

TOOLS= [tool for tool in config['param']]
#Path

path_tool = "../../../"
path_maketool = os.path.join(path_tool,"PipTool")
path_os_minimap=os.path.join(path_maketool,"os-minimap2")
path_sync_minimap=os.path.join(path_maketool,"syncmer_mapping/syncmer_minimap")
path_blend=os.path.join(path_maketool,"blend/bin")
#Function
def two_sample_all(wildcards):
    ret=[]
    a= expand( "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.bam",zip,tool=TOOLS,allow_missing=True)
    for elem in a:
        for specie in config['species']:
            for length in ['5000','350']:
                for er in config['error_rate']:
                    for aparam in config['param'][elem[13:elem.find('_',13)]]['command']:
                        b=elem.replace('{species}',specie)
                        b=b.replace('{error_rate}',er)
                        b=b.replace('{length}',str(length))
                        b=b.replace('{param}',aparam)
                        ret.append(b)
    return ret

def all_sample_default(wildcards):
    ret=[]
    a= expand( "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.bam",zip,tool=TOOLS,param=[config['param'][tool]['command'][0] for tool in TOOLS ],allow_missing=True)
    for elem in a:
        for specie in config['species']:
            for er in config['error_rate']:
                for length in config['length']:
                    b=elem.replace('{species}',specie)
                    b=b.replace('{error_rate}',er)
                    b=b.replace('{length}',str(length))
                    ret.append(b)
    return ret

def all_stat_default(wildcards):
    ret=[]
    a= expand( "stat_tools/{tool}_{species}_{length}_{error_rate}_{param}.stat.txt",zip,tool=TOOLS,param=[config['param'][tool]['command'][0] for tool in TOOLS ],allow_missing=True)
    for elem in a:
        for specie in config['species']:
            for er in config['error_rate']:
                for length in config['length']:
                    b=elem.replace('{species}',specie)
                    b=b.replace('{error_rate}',er)
                    b=b.replace('{length}',str(length))
                    ret.append(b)
    return ret

def all_sample_best(wildcards):
    ret=[]
    a= expand( "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.bam",zip,tool=TOOLS,allow_missing=True)
    for elem in a:
        for specie in config['species']:
            for length in config['length']:
                for er in config['error_rate']:
                    for i in range(len(config['param'][elem[13:elem.find('_',13)]]['command'])):
                            if '*' in config['param'][elem[13:elem.find('_',13)]]['name'][i]:
                                b=elem.replace('{species}',specie)
                                b=b.replace('{error_rate}',er)
                                b=b.replace('{length}',str(length))
                                b=b.replace('{param}',config['param'][elem[13:elem.find('_',13)]]['command'][i])
                                ret.append(b)
    return ret

def getparam(wildcards):
    return wildcards.param.replace("#",' ').replace('ยง','_')

rule run_default:
    input :
       expand( "plots/default/{species}_{length}_{error_rate}_rl{num}.pdf", species=config['species'],length=config['length'],num=['1','2','3','4','5'],error_rate=config['error_rate'])

rule run_params:
    input :
       expand( "plots/params/{tool}/{tool}_{species}_{length}_{error_rate}_bc{num}.pdf",tool=TOOLS, species=config['species'],length=['5000','350'],num=['1','2','3'],error_rate=config['error_rate'])


rule run_best:
    input :
       expand( "plots/best/{species}_{length}_{error_rate}_rl{num}.pdf", species=config['species'],length=config['length'],num=['1','2','3','4','5'],error_rate=config['error_rate'])

rule final_stat:
    input :
        all_stat_default

rule plot_best_command:
    input:
       all_sample_best
    output:
        "plots/best/{species}_{length}_{error_rate}_rl1.pdf",
        "plots/best/{species}_{length}_{error_rate}_rl2.pdf",
        "plots/best/{species}_{length}_{error_rate}_rl3.pdf",
        "plots/best/{species}_{length}_{error_rate}_rl4.pdf",
        "plots/best/{species}_{length}_{error_rate}_rl5.pdf"
    conda:
        "env/plotscript.yaml"
    script:
        'scripts/plot_default.py'


rule plot_default_command:
    input:
       all_sample_default
    output:
        "plots/default/{species}_{length}_{error_rate}_rl1.pdf",
        "plots/default/{species}_{length}_{error_rate}_rl2.pdf",
        "plots/default/{species}_{length}_{error_rate}_rl3.pdf",
        "plots/default/{species}_{length}_{error_rate}_rl4.pdf",
        "plots/default/{species}_{length}_{error_rate}_rl5.pdf"
    conda:
        "env/plotscript.yaml"
    script:
        'scripts/plot_default.py'

rule plot_allstat_command:
    input:
       two_sample_all
    output:
        "plots/params/{tool}/{tool}_{species}_{length}_{error_rate}_bc1.pdf",
        "plots/params/{tool}/{tool}_{species}_{length}_{error_rate}_bc2.pdf",
        "plots/params/{tool}/{tool}_{species}_{length}_{error_rate}_bc3.pdf"
    conda:
        "env/plotscript.yaml"
    script:
        'scripts/plot_params.py'


#-------------converst sam-bam-stat rules-------------------------------
rule sam_stat:
    input:
        "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.bam"
    output:
        "stat_tools/{tool}_{species}_{length}_{error_rate}_{param}.stat.txt"
    conda:
        "env/samtool.yaml"
    shell:
       "samtools stats {input} > {output}"

rule sam_bam:
    input:
        "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.sam"
    output:
        "mapped_reads/{tool}_{species}_{length}_{error_rate}_{param}.bam"
    conda:
        "env/samtool.yaml"
    shell:
       "samtools sort {input} > {output};rm {input}"


#-------------Tools rules-------------------------------

rule all_minimap2:
    input:
        expand("mapped_reads/minimap2_{species}_{length}_{error_rate}_{param}.sam", species=config['species'],length=config['length'],param=config['param']['minimap2']['command'],error_rate=config['error_rate'])

rule minimap2:
    input:
        ref='data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/minimap2_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/minimap2.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/minimap2_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "minimap2 -ax map-ont {params} {input.ref} {input.reads} > {output}"

rule all_graphmap2:
    input:
        expand("mapped_reads/graphmap2_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['graphmap2']['command'],error_rate=config['error_rate'])

rule graphmap2:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/graphmap2_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/graphmap2.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/graphmap2_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "graphmap2 align --auto-rebuild-index -r {input.ref} -d {input.reads} {params} > {output}"


rule all_graphmap:
    input:
        expand("mapped_reads/graphmap_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['graphmap']['command'],error_rate=config['error_rate'])

rule graphmap:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/graphmap_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/graphmap.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/graphmap_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "graphmap align -x sensitive -r {input.ref} -d {input.reads} {params} > {output}"


#rule all_mm2f:
   # input:
     #   expand("mapped_reads/mm2f_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['mm2f']['command'],error_rate=config['error_rate'])

# rule mm2f:
#     input:
#         ref= 'data/ref_{species}.fasta',
#         reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
#     output:
#         "mapped_reads/mm2f_{species}_{length}_{error_rate}_{param}.sam"
#     params:
#         getparam
#     benchmark:
#         "benchmarks/mm2f_{species}_{length}_{error_rate}_{param}.benchmark.txt"
#     shell:
#         "~/bin/mm2-fast/mm2-fast/minimap2 -ax map-ont {input.ref} {input.reads} > {output}"

rule all_blasr:
    input:
        expand("mapped_reads/blasr_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['blasr']['command'],error_rate=config['error_rate'])


rule blasr:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/blasr_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/blasr.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/blasr_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "blasr {input.reads} {input.ref} {params} --sam --out {output}> {output}"

rule all_winnowmap:
    input:
        expand("mapped_reads/winnowmap2_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['winnowmap2']['command'],error_rate=config['error_rate'])


rule winnowmap:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta',
        meril='meril_{species}.txt',
    output:
        "mapped_reads/winnowmap2_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/winnowmap.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/winnowmap2_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "winnowmap -W {input.meril} -ax map-ont {input.ref} {input.reads} > {output}"

rule pre_winnowmap_1:
    input:
        ref= 'data/ref_{species}.fasta',
    output:
       'meril_{species}.txt'
    conda:
        "env/winnowmap.yaml"
    shell:
       """meryl count memory=4 threads=2 k=19 output DB1 {input.ref};
        meryl print greater-than distinct=0.9998 DB1 > {output};
        rm -r DB1"""

rule all_magicblast:
    input:
        expand("mapped_reads/magicblast_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['magicblast']['command'],error_rate=config['error_rate'])


rule magicblast:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta',
    output:
        "mapped_reads/magicblast_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/magicblast.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/magicblast_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "makeblastdb -in {input.ref} -out blastdb/default -parse_seqids -dbtype nucl; magicblast -query {input.reads} -db blastdb/default {params} -out {output};"

rule all_lra:
    input:
        expand("mapped_reads/magicblast_{species}_{length}_{error_rate}_{param}.sam",species=config['species'],length=config['length'],param=config['param']['lra']['command'],error_rate=config['error_rate'])

rule lra:
    input:
        ref= 'data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta',
    output:
        "mapped_reads/lra_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/lra.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/lra_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "lra index -ONT {params} {input.ref};lra align -ONT -t 4 -a {input.ref} {input.reads}  -p s > {output};rm {input.ref}.gli;rm {input.ref}.mmi"

rule all_minimapos:
    input:
        expand("mapped_reads/minimapos_{species}_{length}_{error_rate}_{param}.sam", species=config['species'],length=config['length'],param=config['param']['minimapos']['command'],error_rate=config['error_rate'])

rule minimapos:
    input:
        ref='data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/minimapos_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/minimap2.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/minimapos_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "{path} -ax map-ont {{params}} {{input.ref}} {{input.reads}} > {{output}}".format(path=os.path.join(path_os_minimap,'minimap2'))

rule all_syncminimap:
    input:
        expand("mapped_reads/syncminimap_{species}_{length}_{error_rate}_{param}.sam", species=config['species'],length=config['length'],param=config['param']['syncminimap']['command'],error_rate=config['error_rate'])

rule syncminimap:
    input:
        ref='data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/syncminimap_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/minimap2.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/syncminimap_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "{path} -ax map-ont {{params}} {{input.ref}} {{input.reads}} > {{output}}".format(path=os.path.join(path_sync_minimap,'minimap2'))

rule all_blend:
    input:
        expand("mapped_reads/blend_{species}_{length}_{error_rate}_{param}.sam", species=config['species'],length=config['length'],param=config['param']['blend']['command'],error_rate=config['error_rate'])

rule blend:
    input:
        ref='data/ref_{species}.fasta',
        reads= 'data/samples/{species}_{length}_{error_rate}_reads.fasta'
    output:
        "mapped_reads/blend_{species}_{length}_{error_rate}_{param}.sam"
    conda:
        "env/minimap2.yaml"
    params:
        getparam
    benchmark:
        "benchmarks/blend_{species}_{length}_{error_rate}_{param}.benchmark.txt"
    shell:
        "{path} -ax map-ont {{params}} {{input.ref}} {{input.reads}} > {{output}}".format(path=os.path.join(path_blend,'blend'))

onsuccess:
    shell('if ls data/*.fasta.gmidx 2>/dev/null 1>&2;then rm data/*.fasta.gmidx ;fi;if [[ -d blastdb ]]; then rm blastdb -r ;fi')


